state/protocol/protocol_state/kvstore/models.go

```diff
@@ -13,10 +30,19 @@ package kvstore
 // any software version. Since it is important that the store support managing
 // different model version, this is here so that we can test the implementation
 // with multiple supported KV model versions from the beginning.
-type modelv0 struct{}
+type modelv0 struct {
+	upgradableModel
+}
 
-var _ Reader = new(modelv0)
-var _ API = new(modelv0)
+var _ protocol_state.Reader = new(modelv0)
+var _ protocol_state.API = new(modelv0)
+
+func (model *modelv0) ID() flow.Identifier { return flow.MakeID(model) }
+
+func (model *modelv0) Clone() protocol_state.API {
+	cpy := *model
```

Comment:
⚠️  `VersionUpgrade` is a pointer field of `modelv0`, so this isn't really a deep copy. We should copy pointer fields as well.

Reply:
Yeap, good catch, I've added `VersionUpgrade` after adding `Clone`. I have used a 3rd party library which performs a generic deep copy and added tests: [`2416aa1` (#5513)](https://github.com/onflow/flow-go/pull/5513/commits/2416aa1d03cc8b45dd8c23e66fe31b20d62a6896)